runtime/sema/check_reference_expression.go

```diff
@@ -34,9 +34,16 @@ func (checker *Checker) VisitReferenceExpression(referenceExpression *ast.Refere
 
 	var referenceType *ReferenceType
 	var targetType, referencedType Type
+	var returnType Type
 
 	if !resultType.IsInvalidType() {
 		var ok bool
+		// indexed access to dictionaries require an optional reference as their
+		// target type, so allow one level of optionals
+		optType, optOk := resultType.(*OptionalType)
+		if optOk {
+			resultType = optType.Type
+		}
```

Comment:
maybe leave a blank line after the block
```suggestion
		}

```